### PR TITLE
Move screen created rendering logic to Mobius + Basic Mobius setup

### PR DESCRIPTION
**Pull request by ragunathjawahar**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetController.kt [hunk 1]
```diff
@@ -456,6 +456,8 @@ class BloodPressureEntrySheetController @Inject constructor(
 
     val currentFacilityStream = loggedInUserStream
         .flatMap(facilityRepository::currentFacility)
+        .replay()
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetController.kt]: Why was this added?

Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetController.kt]: Since the stream is being consumed at two different places, `facilityRepository::currentFacility` was being invoked twice. I added this after the characteristic test showed two invocations.

Since this is a behavior we won't be able to replicate in Mobius, I added `replay()` ran the tests, ran the app and ensured the behavior remained the same.

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetController.kt]: Will a `share()` work instead of a `replay().refCount()`?

Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheetController.kt]: `share()` had problems in one of the older migration. Hence avoiding it.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt [hunk 2]
```diff
@@ -72,29 +79,44 @@ class BloodPressureEntrySheet : BottomSheetActivity(), BloodPressureEntryUi {
...
-    setContentView(R.layout.sheet_blood_pressure_entry)
-    TheActivity.component.inject(this)
+  private val viewRenderer = BloodPressureEntryViewRenderer(this)
 
-    bindUiToControllerWithoutDelay(
-        ui = this,
-        events = Observable.mergeArray(
+  private val delegate by unsafeLazy {
+    val defaultModel = when (val openAs = intent.extras!!.getParcelable(KEY_OPEN_AS) as OpenAs) {
+      is New -> BloodPressureEntryModel.newBloodPressureEntry(openAs.patientUuid)
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: I don't understand the point of doing this since `BloodPressureEntryModel` just internally creates a new instance of the same `OpenAs` type anyway? We might as well just pass the `openAs` directly to the constructor of `BloodPressureEntryModel`.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt [hunk 3]
```diff
@@ -108,12 +130,47 @@ class BloodPressureEntrySheet : BottomSheetActivity(), BloodPressureEntryUi {
             dayTextChanges(),
             monthTextChanges(),
             yearTextChanges()
-        ),
+        )
+        .compose(ReportAnalyticsEvents())
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: I just realised that doing this is a little dangerous since `BloodPressureEntrySheetController` also forwards events to analytics. This means that until the controller is completely migrated and deleted, the events will be pushed twice to analytics. This is fine if we are sure that we'll finish the migration in time for the next release, but it's still something to watch out for.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt [hunk 4]
```diff
@@ -108,12 +130,47 @@ class BloodPressureEntrySheet : BottomSheetActivity(), BloodPressureEntryUi {
...
+  }
+
+  override fun onStop() {
+    delegate.stop()
+    super.onStop()
+  }
+
+  override fun onSaveInstanceState(outState: Bundle?) {
+    delegate.onSaveInstanceState(outState)
+    super.onSaveInstanceState(outState)
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Should this be the state that is returned from `MobiusDelegate#onSaveInstanceState`? Maybe we should add a `@CheckReturnValue` annotation so that it isn't missed.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntryViewRenderer.kt [hunk 5]
```diff
@@ -1,11 +1,21 @@
...
 
+import org.simple.clinic.bp.entry.OpenAs.New
 import org.simple.clinic.mobius.ViewRenderer
 
 class BloodPressureEntryViewRenderer(
     val ui: BloodPressureEntryUi
 ) : ViewRenderer<BloodPressureEntryModel> {
   override fun render(model: BloodPressureEntryModel) {
-    // Do nothing!
+    renderOpenAs(model.openAs)
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntryViewRenderer.kt]: Rename method to `setupUi`?

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntryViewRenderer.kt]: I can't think of a good name, this essentially the intention of `set up the ui based on what it was opened for`.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt (hunk 2 above)
Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Makes sense, will fix this.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt (hunk 3 above)
Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Good catch, I missed removing `ReportAnalyticsEvents()` from the controller. Which I can do now. Therefore we could continue shipping the feature even if we are in the middle of refactoring.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntryViewRenderer.kt (hunk 5 above)
Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntryViewRenderer.kt]: Will fix.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt (hunk 3 above)
Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Fixed

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt (hunk 4 above)
Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: I just realized that `onSaveInstanceState(Bundle?)` does not return a value while used inside an `Activity`. We make have to introduce new functions to deal with this case and later split them into two different classes. Will take a stab at this.

Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: I investigated this and it is evident that we need a different delegate for the `Activity`

Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: @vinaysshenoy addressed this feedback. Can you plese review #888 first. So that I can re-request review on this PR. Thanks.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt [hunk 6]
```diff
@@ -99,8 +99,8 @@ class BloodPressureEntrySheet : BottomSheetActivity(), BloodPressureEntryUi {
 
   private val delegate by unsafeLazy {
     val defaultModel = when (val openAs = intent.extras!!.getParcelable(KEY_OPEN_AS) as OpenAs) {
-      is New -> BloodPressureEntryModel.newBloodPressureEntry(openAs.patientUuid)
-      is Update -> BloodPressureEntryModel.updateBloodPressureEntry(openAs.bpUuid)
+      is New -> BloodPressureEntryModel.newBloodPressureEntry(New(openAs.patientUuid))
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: What I meant was that we could just do

```kotlin
val openAs = intent.extras!!.getParcelable(KEY_OPEN_AS) as OpenAs
val defaultModel = BloodPressureEntryModel(openAs)
```

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: The work of deciding the right `OpenAs` is already decided by the previous screen. This screen shouldn't have to worry about it.

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Should we look at capturing the recorded analytics events as well in the characteristics tests, so we know that the events we record have not changed in any way?

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt (hunk 6 above)
Comment (ragunathjawahar) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Since I am migrating 1 feature at a time, I prefer to break into one execution path at once. We can unify this in the final PR.

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Alright.
